Refactoring & Fix: editing group and challenges from user dashboard

Xaz16 opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)


Original https://github.com/HabitRPG/habitica/pull/11357


[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes https://github.com/HabitRPG/habitica/pull/11357#issuecomment-535721541

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

Refactoring edit/delete features
Allow user to edit challenge tasks


[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: d94eca66-c6de-423c-a864-9cb103b9eb77

Xaz16 committed:
Refactoring & Feature: edit/delete group and challenge tasks

- Remove showOption from tasks props
- Pass all needed data to task for understand in mapGetter function which controls we should show
- Improve current solution with edit and delete logic

Xaz16 committed:
Fix: this in template

Xaz16 committed:
Fix & Test: extend tests, fix can Edit/Delete functions

Xaz16 committed:
Merge remote-tracking branch 'upstream/develop' into fix-editing-group-and-challenges-from-user-dashboard

Xaz16 committed:
Fix: allow user edit challenge tasks on dashboard

Xaz16 commented:
Sorry for such late fix
@Alys could you please approve if everything okay? I allowed to edit challenge tasks from user dashboard

paglias commented:
Looks like a client test is failing

Xaz16 committed:
Fix: test case after code change

Xaz16 commented:
@paglias yeah, forgot to update test case according to comment https://github.com/HabitRPG/habitica/pull/11357#issuecomment-535886999

Thank you for noticing :)

paglias committed:
Merge branch 'develop' into fix-editing-group-and-challenges-from-user-dashboard

paglias committed:
fix import path

paglias commented:
Fixed the conflicts generated by some recent client changes

Xaz16 commented:
thank you @paglias, any comments about this PR?

paglias commented:
Sorry about the delay @Xaz16 but we really hadn't time to review PRs this week :(

Xaz16 commented:
@paglias no problem, just want to know that it is still relevant

paglias commented on website/client/src/store/getters/tasks.js:
```diff
@@ -34,12 +34,66 @@ function getTaskColor (task) {
+      case 'challenge':
+        if (challenge && !isTaskFromUserDashboard) {
+          const isUserChallenge = user.id === challenge.leader.id;
+          isUserCanDeleteTask = isUserChallenge || task.challenge.broken;
```
if we're not on the user task board, the challenge is broken but you're not the leader I don't think that the `delete` action should be show as it would be on the challenge's task board anyway. Correct, @Alys ?

Alys replied:
Correct, a challenge participant should never see the delete button on the challenge's task board.

@Xaz16 I don't know if this is relevant since the issue is just about the user's own task board, but if you are changing code that affects the challenge's task board, be aware that moderators/staff have the same edit/delete rights as challenge owners (i.e., a moderator should see a delete button on the challenge's task board, just as the challenge leader does). Here's an example of how to identify moderators along with challenge leaders:
https://github.com/HabitRPG/habitica/blob/1859f431338e16ca9fa2a4245bfe8dd068207315/website/server/models/challenge.js#L80-L83

paglias commented on website/client/src/store/getters/tasks.js:
```diff
@@ -34,12 +34,66 @@ function getTaskColor (task) {
+        if (!isTaskFromUserDashboard) {
+          if (challenge) {
+            const isUserChallenge = user.id === challenge.leader.id;
+            isUserCanEditTask = isUserChallenge || task.challenge.broken;
```
same here, even if the challenge is broken I think only the leader should be able to edit it on the challenge's task board. Also because the `broken` attribute should only exist on the user's copy of the task

paglias commented on website/client/src/components/tasks/task.vue:
```diff
@@ -937,6 +940,27 @@ export default {
+      return !isEmpty(this.task.challenge);
+    },
+    isGroupTask () {
+      return !isEmpty(this.task.group) && (this.task.group.taskId || this.task.group.id);
```
`this.task.group` is always returned form the API even if empty so I think the first part of this check can be removed

paglias reviewed (commented):
left a couple of comments @Xaz16 and sorry for the delay!

Xaz16 committed:
Merge remote-tracking branch 'upstream/develop' into fix-editing-group-and-challenges-from-user-dashboard

Xaz16 committed:
Fix:

- Extend canEdit and canDelete functions with admin role
- Clarify canEdit and canDelete conditions
- Extend test cases

Xaz16 commented:
I fixed all you comments @paglias, and as well extended test cases for my canDelete/canEdit functions... I think for now all should be work as expected

paglias commented:
@Xaz16 sorry again for the delay and thanks for the very well done PR (including tests, ...). This is working well and will go live next week. Welcome to the 3rd tier :)

Xaz16 commented:
@paglias 
Thank you :)
